azure-kusto-data/azure/kusto/data/_models.py

```diff
@@ -146,11 +164,17 @@ def __len__(self):
         return self.rows_count
 
     def __iter__(self):
-        for row in self.rows:
-            yield row
+        for row_index, row in enumerate(self.raw_rows):
+            if self.kusto_result_rows:
+                yield self.kusto_result_rows[row_index]
+            else:
+                kusto_result_row = KustoResultRow(self.columns, row)
+                yield kusto_result_row
 
     def __getitem__(self, key):
-        return self.rows[key]
+        if self.kusto_result_rows:
```

Comment:
Shouldn't something be set if the first condition fails?

Reply:
No, __iter__ and __getitem__ shouldn't make changes

Reply:
wouldn't this cause problems for some use cases?

Reply:
I don't think so. Do I miss something?

Reply:
It seems like accessing the index operator without first "constructing" the result by some other API call, will create results ad-hoc, so if I were to iterate or access them multiple times, it would cause a lot of rows to be created and then disposed (mind my c#)